graphql-java-kickstart/src/main/java/graphql/kickstart/execution/GraphQLInvoker.java

```diff
@@ -0,0 +1,69 @@
+package graphql.kickstart.execution;
+
+import graphql.ExecutionInput;
+import graphql.ExecutionResult;
+import graphql.GraphQL;
+import graphql.kickstart.execution.context.ContextSetting;
+import graphql.kickstart.execution.input.GraphQLBatchedInvocationInput;
+import graphql.kickstart.execution.input.GraphQLInvocationInput;
+import graphql.kickstart.execution.input.GraphQLSingleInvocationInput;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import javax.security.auth.Subject;
+import lombok.RequiredArgsConstructor;
+
+@RequiredArgsConstructor
+public class GraphQLInvoker {
+
+  private final GraphQL graphQL;
+
+  public GraphQLQueryResult query(GraphQLInvocationInput invocationInput) {
+    if (invocationInput instanceof GraphQLSingleInvocationInput) {
+      return GraphQLQueryResult.create(query((GraphQLSingleInvocationInput) invocationInput));
+    }
+    GraphQLBatchedInvocationInput batchedInvocationInput = (GraphQLBatchedInvocationInput) invocationInput;
+    return GraphQLQueryResult.create(query(batchedInvocationInput.getExecutionInputs(), batchedInvocationInput.getContextSetting()));
+  }
+
+  private ExecutionResult query(GraphQLSingleInvocationInput singleInvocationInput) {
+    return executeAsync(singleInvocationInput).join();
+  }
+
+  public CompletableFuture<ExecutionResult> executeAsync(GraphQLSingleInvocationInput invocationInput) {
+    if (Subject.getSubject(AccessController.getContext()) == null && invocationInput.getSubject().isPresent()) {
+      return Subject
+          .doAs(invocationInput.getSubject().get(), (PrivilegedAction<CompletableFuture<ExecutionResult>>) () -> {
+            try {
+              return query(invocationInput.getExecutionInput());
+            } catch (Exception e) {
+              throw new RuntimeException(e);
+            }
+          });
+    }
+
+    return query(invocationInput.getExecutionInput());
+  }
+
+  private CompletableFuture<ExecutionResult> query(ExecutionInput executionInput) {
+    return graphQL.executeAsync(executionInput);
+  }
+
+  private List<ExecutionResult> query(List<GraphQLSingleInvocationInput> batchedInvocationInput,
+      ContextSetting contextSetting) {
+//    List<ExecutionInput> executionIds = batchedInvocationInput.stream()
+//        .map(GraphQLSingleInvocationInput::getExecutionInput)
+//        .collect(Collectors.toList());
+//    Supplier<Instrumentation> configuredInstrumentation = contextSetting
```

Comment:
@jmccaull I've created a separate `GraphQLBuilder` to create the `GraphQL` object to run the queries against. I thought it would improve performance if you'd just create that once instead of on every query. This is the only part though that I couldn't easily refactor. Any thoughts on this?

I'm also working on a webflux starter, which means we'd have to support a `CompletableFuture` result for batches too.

Reply:
It has been a while since we did any profiling, I don't remember this showing up at all as a hotspot or anything. The documentation on the GraphQl object states "Building this object is very cheap and can be done on each execution if necessary...". That being said reusing it would be nice, cheap is relative and documentation gets outdated. It would certainly remove a lot of complexity from invoking a query too.
What I see making this hard to refactor is the instrumentation - I wrote it around the object being recreated for each execution while trying to reuse as much of the original dispatch code as possible, that seems like a mistake from this perspective. The behavior for each context settings is so different would it make sense to create a separate invoker for each? It seems like the request scoped logic would need a handle not only to the GraphQl object but to the instrumentation as well to set it up for each request. At that point too it is probably worth revisiting the tracking approach abstraction, maybe making two distinct instrumentations.
Let me know if you want any help with the lift, etc. It seems like a fairly large one

Reply:
I decided to create it on the fly anyway, but still refactoring it a bit.